Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_LRU_REFERENCE_DELEGATE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_LRU_REFERENCE_DELEGATE_H_
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/local/lru_garbage_collector.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_delegate.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class ListenSequence;
+class MemoryPersistence;
+class Sizer;
+
+/**
+ * Provides the LRU GC implementation for memory persistence.
+ */
+class MemoryLruReferenceDelegate : public LruDelegate {
+ public:
+  MemoryLruReferenceDelegate(MemoryPersistence* persistence,
+                             LruParams lruParams,
+                             std::unique_ptr<Sizer> sizer);
+
+  bool IsPinnedAtSequenceNumber(model::ListenSequenceNumber upperBound,
+                                const model::DocumentKey& key);
+
+  // MARK: ReferenceDelegate overrides
+
+  model::ListenSequenceNumber current_sequence_number() override;
+
+  void AddInMemoryPins(ReferenceSet* set) override;
+
+  void AddReference(const model::DocumentKey& key) override;
+
+  void RemoveReference(const model::DocumentKey& key) override;
+
+  void RemoveMutationReference(const model::DocumentKey& key) override;
+
+  void RemoveTarget(const QueryData& queryData) override;
+
+  void UpdateLimboDocument(const model::DocumentKey& key) override;
+
+  void OnTransactionStarted(absl::string_view label) override;
+
+  void OnTransactionCommitted() override;
+
+  // MARK: LruDelegate overrides
+
+  LruGarbageCollector* garbage_collector() override;
+
+  int64_t CalculateByteSize() override;
+
+  size_t GetSequenceNumberCount() override;
+
+  void EnumerateTargets(const TargetCallback& callback) override;
+
```

Comment:
Same here: I think `Calculate` and `Get` could be grouped together, then both `Enumerate`s, and finally both `Remove`s.